docs/enhancements/automatic-agent-classification-labels.md

```diff
@@ -0,0 +1,112 @@
+
+### Goals
+
+- Provide a flexible and easy-to-use API for defining desired labels and conditions.
+- Extend the Infrastructure Operator to maintain the desired labels based on the specified conditions with minimal impact on scale.
+
+### Non-Goals
+
+- Ship pre-defined labels and conditions.
+
+## Proposal
+
+### User Stories
+
+#### Story 1
+
+As a Cluster Creator, I want the system to automatically label Agents based on specific conditions so that I can more easily select Agents for cluster membership.	
+
+## Design Details [optional]
+
+We will create a new AgentClassification namespace-scoped CRD that defines the desired label key-value and a query to be evaluated according to the Agent’s inventory.
+
+### AgentClassification CRD Definition
+
+```
+const (
+	QueryValidCondition conditionsv1.ConditionType = "QueryValid"
+	QueryValidReason    string                     = "QueryIsValid"
+	QueryNotValidReason string                     = "QueryIsNotValid"
+)
+
+// AgentClassificationSpec defines the desired state of AgentClassification
+type AgentClassificationSpec struct {
+	// LabelKey specifies the label key to apply to matched Agents
+	//
+	// +immutable
+	LabelKey string `json:"labelKey"`
+
+	// LabelValue specifies the label value to apply to matched Agents
+	//
+	// +immutable
+	LabelValue string `json:"labelValue"`
+
+	// A query in jq format that will be invoked on each Agent's inventory.
+	// The query should return a boolean. The operator will apply the label
+	// to any Agent for which "true" is returned.
+	Query string `json:"query"`
+}
+
+// AgentClassificationStatus defines the observed state of AgentClassification
+type AgentClassificationStatus struct {
+	// MatchedCount shows how many Agents currently match the classification
+	MatchedCount int `json:"matchedCount,omitempty"`
+
+	Conditions []conditionsv1.Condition `json:"conditions,omitempty"`
+}
+```
+
+We will base the expression definition on the [gojq](https://github.com/itchyny/gojq) library, which supports jq queries in Go.  The query will be run on each Agent's inventory.  If the query returns true, then the specified label will be applied.
+
+For example,
+The query for label "size:medium" might be: ".cpu.count == 2 and .memory.PhysicalBytes >= 4294967296 and .memory.PhysicalBytes < 8589934592"
+The query for label "storage:large" might be: "[.disks[] | select(.sizeBytes > 1073741824000)] | length > 5"
+
+### Operator changes
+
+When an Agent's inventory is updated, the Agent controller will now need to list all AgentClassification CRs in the namespace and update the Agent's labels accordingly.
+
+A new controller will watch AgentClassification CRs.  Upon creation/update, it will list all Agents in the namespace and apply the relevant labels.  It will also ensure that the specified label key is unique in that namespace.  Upon deletion, it will delete the label from all Agents.
+
+### Risks and Mitigations
```

Comment:
As a user looking at a particular agent, how can I tell apart these two situations :
1. Agent didn't match any labels
2. The controller didn't label the agent yet

Would it be possible to add a meta label/annotation to all agents indicating when they were last processed by the labeling controller? It would eliminate this ambiguity and provide consumers with a way to ensure no races occur when processing those labels (after creating a new AgentClassification, a consumer wouldn't have to wait an arbitrary amount of time while the controller is processing possibly thousands of agents - they could use this timestamp indicator to ensure a particular agent has been classified)

Reply:
Alternatively, if it's possible from a technical perspective (I'm not sure it is), it might be better to give each AgentClassification a "done processing" status.

Not sure how to define such thing though, since agents can be created / deleted all the time